HACKING-URL.md

```diff
@@ -0,0 +1,319 @@
+
+### Sidebar display
+
+:warning:
+The MVC pattern is bent here as the names displayed for hyperlinks are coded in
+the entity model layer. Plans are to address this issue tracked with the ticket
+[MBS-10605](https://tickets.metabrainz.org/browse/MBS-10605).
+
+For a domain to be displayed on the sidebar, you’ll need to create an URL model
+in [`lib/MusicBrainz/Server/Entity/URL/`](lib/MusicBrainz/Server/Entity/URL/).
+Base your model on any of the existing ones that are known to go by the same
+types of relationship, and name it after the site in question.
+
+The `sidebar_name` method most often returns a simple untranslated string,
+the name of the site. You can also return a translated string, most often when:
+* The name of the site has official localizations, for example `l(Niconico)`;
+* The site is mainly about streaming, for example `l('Stream at YouTube Music')`;
+* The site is about scores, even for just some paths, for example the more
+  complex method in [IMSLP](lib/MusicBrainz/Server/Entity/URL/IMSLP.pm).
+
+:warning:
+The two latter cases are temporary fallbacks to the lack of display layout;
+Instead plans are to smartly group external links by relationship type;
+See [MBS-10605](https://tickets.metabrainz.org/browse/MBS-10605) again.
+
+For your shiny new URL model to be actually used when loading URLs, you also
+need to add the mapping to `%URL_SPECIALIZATIONS` in
+[`MusicBrainz::Server::Data::URL`](lib/MusicBrainz/Server/Data/URL.pm).
+
+### Rendered scheme override
+
+There are mainly two reasons for overriding URL scheme (`http`/`https`) on
+display:
+
+If the external site supports both `http` and `https` schemes,
+then override the method `url_is_scheme_independent` with `{ 1 }`
+in the corresponding `Entity` model added for sidebar display.
+That way, the URL scheme will match the scheme of the MusicBrainz Server
+instance (`https` for `musicbrainz.org`, usually `http` for mirrors).
+See [InternetArchive](lib/MusicBrainz/Server/Entity/URL/InternetArchive.pm) for example.
+
+If the external site dropped support for `http` scheme or redirect from `http`
+to `https`, and its URLs stored in the MusicBrainz database still use `http`
+(for example to keep following the permalink format),
+then override the method `href_url` to make the appropriate change;
+That way, the URL scheme will be systematically overridden accordingly.
+See [VIAF](lib/MusicBrainz/Server/Entity/URL/VIAF.pm) for example.
+
+### Custom in-page display
+
+All external links are shown under the “Relationships” tab of entity pages.
+
+You will notice that some very rare websites (Amazon, IMSLP, VIAF, Wikidata...)
+have a custom `pretty_name` for URL display here. These websites actually have
+dedicated relationship types which names are used as label ahead of the URL so
+that the identifier contained in the URL can be shown instead of the full URL.
+
+So it is pointless to have a custom `pretty_name` for other websites.
```

Comment:
Maybe it should instead be moved after the next paragraph that is discussed in the above conversation.

Reply:
Maybe? Dunno. In any case, I can perfectly well imagine having another URL type where we want to display just the ID, so I don't think we should say it will always be pointless.

Reply:
I’m not sure to follow your comment. What do you mean with “another URL type”?

Should it be clarified that “other websites” refer to sites that do not “have dedicated relationship types”?

Reply:
Oh, so what you meant is "you shouldn't use this for stuff that does not have a dedicated relationship"? Yeah, in that case that seems sensible :)